Copula 7254 rebased

josef-pkt opened the pull request:
amending last commit and getting unit tests to pass

josef-pkt commented:
unit test only fail with missing matplotlib
and a few pep-8 complaints

josef-pkt commented:
coverage report codecov looks screwed up.
None of my unit test already in main are run anymore according to codecov
 (but I'm pretty sure thwy were run locally on my computer.)

pep8speaks commented:
Hello @josef-pkt! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`statsmodels/distributions/copula/archimedean.py`](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/archimedean.py):

> [Line 245:60](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/archimedean.py#L245): [E131](https://duckduckgo.com/?q=pep8%20E131) continuation line unaligned for hanging indent



* In the file [`statsmodels/distributions/copula/elliptical.py`](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/elliptical.py):

> [Line 84:30](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/elliptical.py#L84): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file
> [Line 84:30](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/elliptical.py#L84): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file
> [Line 84:30](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/elliptical.py#L84): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`statsmodels/distributions/copula/other_copulas.py`](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/other_copulas.py):

> [Line 39:70](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/other_copulas.py#L39): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`statsmodels/distributions/copula/tests/test_copula.py`](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/tests/test_copula.py):

> [Line 379:73](https://github.com/statsmodels/statsmodels/blob/0b1ed3530c11f3c0387f3302c30c409b55ae8275/statsmodels/distributions/copula/tests/test_copula.py#L379): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

josef-pkt commented:
error 1E   AttributeError: module 'scipy.stats' has no attribute 'multivariate_t'1
in test using scipy 1.3.3

josef-pkt commented on statsmodels/distributions/copula/elliptical.py:
```diff
@@ -6,3 +6,79 @@
+        if cov is None:
+            cov = [[1., 0.], [0., 1.]]
+        self.density = stats.t(df=df)
+        self.mv_density = stats.multivariate_t(shape=cov, df=df)
```
misleading names:
those are distribution instances and not densities/pdf

josef-pkt commented:
something weird

why is pytest running the `__init__` of a class that does not start or end in `Test`?
class name is `_estStudentCopula`, super class also doesn't have test in name anymore

```
____ ERROR collecting statsmodels/distributions/copula/tests/test_copula.py ____
statsmodels/distributions/copula/tests/test_copula.py:342: in <module>
    class _estStudentCopula(CheckCopula):
statsmodels/distributions/copula/tests/test_copula.py:343: in _estStudentCopula
    copula = StudentCopula(cov=[[1., 0.8], [0.8, 1.]], df=2)
statsmodels/distributions/copula/elliptical.py:68: in __init__
    self.mv_density = stats.multivariate_t(shape=cov, df=df)
E   AttributeError: module 'scipy.stats' has no attribute 'multivariate_t'
____ ERROR collecting statsmodels/distributions/copula/tests/test_copula.py ____
statsmodels/distributions/copula/tests/test_copula.py:342: in <module>
    class _estStudentCopula(CheckCopula):
statsmodels/distributions/copula/tests/test_copula.py:343: in _estStudentCopula
    copula = StudentCopula(cov=[[1., 0.8], [0.8, 1.]], df=2)
statsmodels/distributions/copula/elliptical.py:68: in __init__
    self.mv_density = stats.multivariate_t(shape=cov, df=df)
E   AttributeError: module 'scipy.stats' has no attribute 'multivariate_t'
```

trying again, this time commented out the test class
amended and force pushed

josef-pkt commented:
Ok that was it.

pytest didn't run the full test suite. It skipped more than just the (non-test) class that had an error, i.e. stopped collecting. pytest :(
Now test coverage is back up to 75%, 
archimedean, elliptical and others are mostly covered, ev is unchanged

tupui commented:
Just seeing this, shall I close #7254 then? What's the plan here @josef-pkt? Let me know if I can still be of help, even just for reviewing it's fine.

josef-pkt commented:
We need a copy of multivariate_t for compatibility with scipy < 1.6
Currently I use the copied version unconditionally to check behavior across versions.
I will make it conditional on scipy <1.6 later.

josef-pkt commented:
now all green including style check
unit test coverage looks ok (> 80%) except for copulas.py (old code, plot method)

trying out:

plot requires dimension d=2
GaussianCopula has d=inf

--- 

ValueError                                Traceback (most recent call last)
<ipython-input-7-f429d75ae986> in <module>
      1 cop = GaussianCopula(cov=[[1., 0.8], [0.8, 1.]])
----> 2 _ = cop.plot_pdf(10)

m:\josef_new\eclipse_ws\statsmodels\statsmodels_py37_main\statsmodels\statsmodels\distributions\copula\copulas.py in plot_pdf(self, ticks_nbr, ax)
    414         from matplotlib import pyplot as plt
    415         if self.d != 2:
--> 416             raise ValueError("Can only plot 2-dimensional Copula.")
    417 
    418         n_samples = 100

ValueError: Can only plot 2-dimensional Copula.

---

also ZeroDivision Warning in plot for GaussianCopula

josef-pkt commented:
StudentCopula has default df=1
Either we don't set a default or we set it too a larger value.

I got a bit surprised because the contourplot for T copula with gaussian margins looks almost like an X (tilted cross)

with df=3 or 4, the contour/scatter plot looks more like a trapezoidal ellipsoid

josef-pkt commented:
naming/terminology

elliptical distribution cov/corr/scatter/shape matrix   What do we call it?
in gaussian the matrix is pearson correlation 
in mv-t the matrix does not correspond to cov of distribution, scipy uses `shape` matrix.

correction: AFAICS, standard mv-t shape matrix is the same as the correlation matrix. Variance and cov depend on df and scale, and standard t, mv-t (scale=1) does not have unit variance, but df term should cancel when we compute pearson correlation.

-> use `corr` or `pearson_corr` for shape matrix in Gaussian and t copulas, instead of current cov.

josef-pkt commented:
unit tests: unnecessary repeated sampling `sample = self.copula.random(10000)`

josef-pkt commented:
two unrelated errors in azuere testing
windows 3.8 fails to install `Building wheel for pywinpty (PEP 517): finished with status 'error'`
linux pre has a test error which looks like scipy.spatial pdist changes

this PR is still green, including style check

josef-pkt commented on statsmodels/distributions/copula/copulas.py:
```diff
@@ -177,6 +182,40 @@ def __init__(self, marginals, copula, copargs=()):
+            # this means marginals are independents
+            sample = rng.random((n, self.d))
+        else:
+            sample = self.copula.random(n, random_state=random_state)
```
why does this not use `rng`?
what is return of check_random_state

tupui replied:
It's a typo and should read `sample = self.copula.random(n, random_state=rng)`. `check_random_state` returns an instance of `RandomState` or `Generator` (only if a generator was passed).

josef-pkt commented:
`_rankdata_no_ties` should be 1- based, currently it's zero based.  BUG
I need to check options for "plotting positions" in usage of ranks for copulas

our current univariate ECDF uses `y = np.linspace(1./nobs,1,nobs)` i.e. no slack, like `arange(1, nobs + 1) / nobs
For distributions we don't have in general fixed finite support bounds
For copulas, we require [0, 1]. and use plotting position to stay away from boundary or only from lower boundary.
( I guess it depends on whether we want points or bins #7454 )

`plotting_pos` function is in  `statsmodels.graphics.gofplots`
aside: class ProbPlot only uses `a` but not `b` parameter: `plotting_pos(self.nobs, self.a)`
```
b = a if b is None else b
return (np.arange(1.0, nobs + 1) - a) / (nobs + 1 - a - b)
```
a=b=0 is "weibull"  1 ... n / (n + 1)
a=b=0.5 is "hazen" ({1 ... n} - 0.5)  / n  
names in copBasic docs

josef-pkt commented:
maybe I shouldn't ignore ties for ecdf, rankdata
example elnino monthly temperature `dataset` . Data is rounded to two decimals and has ties
first column is year, others are temp by months
```
(np.diff(np.sort(dta.data.values, axis=0), axis=0) == 0).sum(0)
array([0, 8, 6, 5, 6, 4, 6, 9, 6, 1, 5, 5, 5])
```

I started unit test against R copBasic with this, but arbitrary tie breaking looks fragile and might agree only as long as same sorting algorithm is used across packages. (For Hosmer-Lemeshow grouping I also had tie breaking differences by sort algorithm or tie breaking rules.)

I think brute force ecdf on non-ranked data is unique. Other cases might not produce correct ecdf with ties. 
That would require rankdata with ties in ranking, and no sequential search (my speedup for bivariate case).

josef-pkt commented:
design: make dimension a `__init__` keyword.

We need copula parameters as method arguments at least for some methods like logpdf/loglike. But we can treat the dimension of the model as fixed. In some cases I kept it currently implicit, e.g. determining dimension from values of random variable u.

One reason to do that is that we might precompute some intermediate results for fixed dimension.

e.g. Hofert et al have explicit formulas for multi-dimensional pdf/logpdf of 5 archimedean copula families. Gumbel and Joe have a finite sum term that is indepenend of `u`, however not independent of copula parameter and dimension.

Frank and AMH are messier because they have infinite sum, polylogarithm that doesn't exist in scipy.special. That sum also depends on `u`.

Hofert, Marius, Martin Mächler, and Alexander J. McNeil. 2012. “Likelihood Inference for Archimedean Copulas in High Dimensions under Known Margins.” Journal of Multivariate Analysis, Special Issue on Copula Modeling and Dependence, 110 (September): 133–50. https://doi.org/10.1016/j.jmva.2012.02.019.

josef-pkt commented:
last commit is WIP, committing what I left over 2 or 3 weeks ago.

(got side tracked again)

josef-pkt commented:
syntax error in last commit

josef-pkt reviewed (commented):
I had forgotten to submit those

tupui commented:
Hi, just checking in. Do you have an ETA for this one? Would be happy to use it in the next release. Let me know if I can have be of help.

josef-pkt commented:
I will get back to this after merging Beta regression.

I will merge this PR with missing pieces, but try to have what's available sufficiently unit tested.

tupui commented:
Glad to hear!

josef-pkt commented:
rebased on current main

josef-pkt commented:
it looks like the last commit was unfinished refactoring adding random_state. has syntax error
`tau_simulated(self, nobs=1024, random_state=):`

josef-pkt commented on statsmodels/distributions/copula/copulas.py:
```diff
@@ -261,3 +300,229 @@ def logpdf(self, y, args=None):
+    """
+
+    def __init__(self, d):
+        self.d = d
```
no one letter names
k_dim or k_vars

josef-pkt commented on statsmodels/distributions/copula/copulas.py:
```diff
@@ -177,6 +182,40 @@ def __init__(self, marginals, copula, copargs=()):
+        rng = check_random_state(random_state)
+        if self.copula is None:
+            # this means marginals are independents
+            sample = rng.random((nobs, self.d))
```
this should go through independence copula, not here in base class

josef-pkt commented:
insuffient unit test coverage in several modules https://codecov.io/gh/statsmodels/statsmodels/tree/copula_7254_rebased/statsmodels/distributions/copula

`copulas.py` still contains old code, not covered
`other_copulas.rvs_kernel` added in last commit has not unit tests
`archimedean.py` missing test coverage mainly for tau related methods
`elliptical.py` tau, rho related methods and tail
`depfunc_ev.py` mainly no coverage for `_check_args`
`extreme_value` 100% coverage
`is_completly_monotonic` only method `is_completly_monotonic` is not covered

`distributions.tools.py` no coverage for `_ecdf_mv` function

josef-pkt commented:
based on my latest notebook (before the break) I was working on ecdf and beta kernel.
dataset elnino using two months would work nicely, except that there are ties that ecdf is not designed to handle.

AFAICS, I didn't manage to replicate this with package `library(copBasic)` with something like
```
> fakeV <- lmomco::pp(en2[,2], sort=FALSE) # Weibull plotting position i/(n+1)
> uv <- data.frame(U=fakeU, V=fakeV); # our U-statistics
> empcop <- EMPIRcopdf(para=uv)
```

en2 is elnino with april and august in the two columns

josef-pkt commented:
@tupui I will merge this today.
It still needs work before the release, but I will be travelling and sightseeing for the next two weeks.

You could add missing pieces if you have time.

One part that needs to be cleaned up and added to is the kendall `tau` support. We need either analytical functions or numerical integration for those. I added some analytical tau function, to elliptical, gumbel and clayton.
Also, we need the inverse function for tau to get the copula correlation from tau. I don't know how much we have there.

extreme value copulas will not have a tau or inverse tau directly, because those all have more than one parameter.
I don't know yet how to support the symmetric special cases for the extreme value copulas.

currently only Frank has conditional distribution of the second rv given the first rv: `cdfcond_2g1`, `ppfcond_2g1`

We might be able to support main parts pdf, cdf for more than two dimension in some Archimedean copulas, maybe gumbel or clayton ?

josef-pkt commented:
I forgot to push my latest changes using amend

ArchimedeanCopula.cdf and other methods have two docstrings. I fixed it but didn't push it before merging.

tupui commented:
Thanks for merging @josef-pkt

I will see if I have some time for tau.
Also before the release it would need some user friendly documentation on how to use this. I can try to propose something. Where should this go?

josef-pkt commented:
Some overview description can go into the module docs. I think GLM has the most of that.

examples for usage go into notebooks.
Notebooks that are added to the repository need to be empty in the PR, so we don't increase the size of a checkout with machine generated data, plots, ...
We can use a gist for developing a notebook so we can see the output.

I have several notebooks that I started out as usage examples, but then become development notebooks when I needed to try out new features.

tupui commented:
Something else, would it be ok if I added support for QMC? There would not be much to do. We are using `rng.random` all around and we would just need to somehow allow `QMCEngine.random`, I could write a small wrapper for it.

Either way the current `check_random_state` is not the best solution as it's a private method we should not rely too much on. So I could write a new one for `statsmodels` if you want, with the requirements you want. I can also just copy it and add support for passing along `QMCEngine`.

josef-pkt commented:
Yes to both
Random state helper function should go into statsmodels.tools, I think

Qmc would be good to add, but many copula rvs are based on mixtures, and those will not have proper properties.
We need conditional ppf to support qmc rvs
